### ISSUE TITLE
[Clang] Control-flow analysis around expansion statements could be better

### ISSUE THREAD

**Issue by Sirraide** [llvm/llvm-project#211939]
Concretely, consider (https://godbolt.org/z/e75Kx1WMr)
```c++
int f() {
    template for (int x : {1}) {
        return 4;
    }
}
```
which currently results in a warning
```console
<source>:5:1: warning: non-void function does not return a value [-Wreturn-type]
    5 | }
      | ^
```
despite the fact that we _do_ in fact always return here, as the expansion statement effectively expands to
```c++
int f() {
    {
        {
            int x = 1;
            return 4;
        }
    }
}
```
which we don’t warn about. Presumably, the fix is that we should probably be walking the `CXXExpansionStmtInstantiation` of an expansion statement rather than the pattern.

**Comment by efriedma-quic** [comment]
Another variation which I think is basically the same thing:

```c++
int f() {
  template for (constexpr auto x : {1}) {
    goto A;
  }
  struct A { A() { } } x;
  A: return 0;
}
```
